### ISSUE TITLE
investigate if modules need to be deserialized on every execution of lambda function.

### ISSUE THREAD

**Issue by pberkland** [EclairJS/eclairjs-nashorn#269]
currently every time a lambda function is executed javaToJs is executed on the module parameters.
It seems like for module parameters, it only needs to be done one time.  The javascript function source is only complied (eval) the first time a lambda function is executed.

**Comment by billreed63** [comment]
This does not seem to have a significant impact
The code as it is loading the movie-lens large data set with this code

```
  var complete_ratings_raw_data = sc.textFile("examples/data/mllib/ml-latest/ratings.csv");
   var start = new Date().getTime();

   var complete_ratings_data = complete_ratings_raw_data
       .filter(function(line) {
            // filters out the header
            return line != "userId,movieId,rating,timestamp"; //complete_ratings_raw_data_header;
        })

        .map(function(line, Rating) {
            var tokens = line.split(",");
            // tokenes are userId,movieId,rating,timestamp
            return new Rating(tokens[0],tokens[1],tokens[2]);
            return line;
        },[Rating]).cache()


    print("There are recommendations in the complete dataset:  " + complete_ratings_data.count());

    var end = new Date().getTime();
    var time = end - start;
    print('Execution time: ' + time + " milliseconds");
```

Has a run time of

```
There are recommendations in the complete dataset:  22884377
Execution time: 45746 milliseconds
```

Changing `JSFunction` to only serializes once 

```
public class JSFunctionTest3 implements Function {
    private String func = null;
    private Object args[] = null;
    private Object ser_args = null;
    private Object fn = null;

    public JSFunctionTest3(String func, Object[] o) {
        this.func = func;
        this.args = o;
    }

    @SuppressWarnings({ "null", "unchecked" })
    @Override
    public Object call(Object o) throws Exception {
        ScriptEngine e =  NashornEngineSingleton.getEngine();
        if (this.fn == null) {
            this.fn = e.eval(func);
        }
        Invocable invocable = (Invocable) e;
        if(this.ser_args == null) {
            this.ser_args = invocable.invokeFunction("Utils_serializeBindArgs", this.args);
        }
        Object x = invocable.invokeFunction("Utils_serializeBindArgs", o);

        Object params[] = {this.fn, x, this.ser_args};

        /*if (this.args != null && this.args.length > 0 ) {
            params = ArrayUtils.addAll(params, this.args);
        }*/

        Object ret = invocable.invokeFunction("Utils_invoke2", params);

        return ret;
    }
}
```

```
function Utils_serializeBindArgs() {
    //var args = [];
    var a = Array.prototype.slice.call(arguments);
    var args = (arguments.length > 0)
        ? a.slice(0).map(function (arg) {
        return Serialize.javaToJs(arg);
    })
        : [];
    return args;
}

function Utils_invoke2(func) {
    var fn = func;
    var a = Array.prototype.slice.call(arguments);
   /* var args = (arguments.length > 1)
        ? a.slice(1).map(function (arg) {
        return arg;
    })
        : [];;*/
    var args = a[1].concat(a[2])
    var ret = null;
    try {
        ret = Serialize.jsToJava(fn.apply(this, args));
    } catch (err) {
        print("error invoking function");
        print(func);
        print(err);
        throw err;
    }

    return ret;
};
```

add updating `RDD.map` To use `JSFunctionTest3`
Results:

```
There are recommendations in the complete dataset:  22884377
Execution time: 44564 milliseconds
```

**Comment by billreed63** [comment]
Running the LAMBDA functions in Nashorn has a cost, running code that just loads a large dataset and filters the dataset removing a string.

```
    var obj = complete_ratings_raw_data.getJavaObject();

    var start = new Date().getTime();
    var complete_ratings_data = obj.filter(new org.eclairjs.nashorn.JSFunctionTest());


    print("There are recommendations in the complete dataset:  " + complete_ratings_data.count());

    var end = new Date().getTime();
    var time = end - start;
    print('Execution time: ' + time + " milliseconds");
```

Running the filter in Java with 

```
package org.eclairjs.nashorn;

import org.apache.commons.lang.ArrayUtils;
import org.apache.spark.api.java.function.Function;

import javax.script.Invocable;
import javax.script.ScriptEngine;

public class JSFunctionTest implements Function {


    public JSFunctionTest() {

    }

    @SuppressWarnings({ "null", "unchecked" })
    @Override
    public Object call(Object l) {
        String line = (String) l;
        if (line.equals("userId,movieId,rating,timestamp")) {
            return false;
        } else {
            return true;
        }
    }
}
```

gives us a time of

```
There are recommendations in the complete dataset:  22884377
Execution time: 2379 milliseconds
```

Changing the LAMDA to run in nashorn using

```
import org.apache.commons.lang.ArrayUtils;
import org.apache.spark.api.java.function.Function;

import javax.script.Invocable;
import javax.script.ScriptEngine;

public class JSFunctionTest2 implements Function {

    private Object fn = null;

    public JSFunctionTest2() {

    }

    @SuppressWarnings({ "null", "unchecked" })
    @Override
    public Object call(Object l) throws Exception {
        ScriptEngine e =  NashornEngineSingleton.getEngine();
        if (this.fn == null) {
            String func = "function myTestFunc(line) { return line != \"userId,movieId,rating,timestamp\";}";
            this.fn = e.eval(func);
        }
        Invocable invocable = (Invocable) e;

        Object params[] = {this.fn, l};


        Object ret = invocable.invokeFunction("myTestFunc", params);

        return ret;
    }
}
```

gives us a time of:

```
There are recommendations in the complete dataset:  22884378
Execution time: 8372 milliseconds
```

just using Nashorn to run the equivalent JavaScript code without our serialization cost us 60 milliseconds

**Comment by pberkland** [comment]
Removing the module reference from the large dataset causes time to go from 44 sec to 30 sec

>  .map(function(line, Rating) {
>             var tokens = line.split(",");
>             // tokenes are userId,movieId,rating,timestamp
>             return new Rating(tokens[0],tokens[1],tokens[2]);
>         }, [Rating]).cache()

changed to 

> .map(function(line) {
>             var tokens = line.split(",");
>             // tokenes are userId,movieId,rating,timestamp
>             return tokens.join();
>         }).cache()

**Comment by pberkland** [comment]
fixed by commit a8c7d71cde951989fe454b503d109fc633916c4f
